pageserver/src/tenant.rs

```diff
@@ -1393,13 +1451,107 @@ impl Tenant {
             timeline
         };
 
+        // multiple concurrent requests could had "transitioned" a timeline to stopping state,
+        // because repeat transitions to same state are allowed. because the delete_timeline will
+        // need to synchronize and cannot be original semi-lockless algorithm for "upload
+        // indexpart" part, we create a single task to delete the timeline.
+
+        let mut rx = {
+            let mut g = timeline.delete_self.lock().await;
+
+            // see if we should join any existing attempt
+            let maybe_rx = if let Some(maybe_done) = g.as_ref() {
+                use timeline::MaybeDone;
+
+                match maybe_done {
+                    MaybeDone::Done(Ok(())) => return Ok(()),
+                    MaybeDone::Done(Err(e)) if e.is_permanent() => {
+                        return Err(DeleteTimelineError::from(e))
+                    }
+                    MaybeDone::Pending(rx) => {
+                        // important: upgrading and later resubscription happens while holding the lock
+                        rx.upgrade()
```

Comment:
Something I noted.

It's more important that no one holds an upgraded `Weak` around except for the duration of the locked section here, which implies these "upgrading and resubscription".

Something I don't like that the `witnessing None while failing to upgrade is "silently" retryable task panicked`. This could be solved with number of ways:

- enforce that we send always send value on drop when panicking
  - this would mean that the only reason for finding a non-upgradeable `Weak` would then be what...? Perhaps it should never happen.
  - drop guard cannot set the value behind the lock so probably not a great idea
    - or just means that we cannot have the async lock
- seems easier to go with the current non-upgradeable weak => panic has happened => should be retried
  - very difficult to test to make sure there's no chance of extra closed messages, which are currently turned into InternalError's

This sort of ties into the refactoring of the MaybeDone into reusable thing. Without doing that, we cannot really hope to unit test it. What testing would really require is all combinations 2-3 tasks can have for await points, otherwise only a positive test case could be written which proves nothing.

Reply:
What does the `Weak` currently buy us: retrying because without looping, and it seems natural because we can have other transient errors.

If there was a loop, and we'd restart it upon seeing `Closed` then that would similarly work, but raise other questions on boundedness and thundering herds attemping to restart.

Reply:
Any loop will have to be outside SharedRetried in #4194; resolving.